File: Android/app/src/main/java/uk/ewancroft/inkwell/util/TipPromptManager.kt
package uk.ewancroft.inkwell.util

import android.content.Context
import android.content.SharedPreferences
import uk.ewancroft.inkwell.shared.policy.TipPromptPolicy

object TipPromptManager {

    private const val PREFS_NAME = "inkwell_tip_prompt"
    private const val KEY_LAUNCH_COUNT = "launch_count"
    private const val KEY_LAST_SHOWN = "last_shown_date"

    fun recordLaunch(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val current = prefs.getInt(KEY_LAUNCH_COUNT, 0)
        prefs.edit().putInt(KEY_LAUNCH_COUNT, current + 1).apply()
    }

    fun shouldShowTip(context: Context): Boolean {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val count = prefs.getInt(KEY_LAUNCH_COUNT, 0)
        val lastShown = prefs.getLong(KEY_LAST_SHOWN, -1L)

        return TipPromptPolicy.shouldShowTip(
            launchCount = count,
            lastShownEpochMillis = lastShown,
            nowEpochMillis = System.currentTimeMillis(),
        )
    }

    fun markShown(context: Context) {
        val prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        prefs.edit().putLong(KEY_LAST_SHOWN, System.currentTimeMillis()).apply()
    }
}

File: iOS/Inkwell/Features/TipPromptManager.swift
//
//  TipPromptManager.swift
//  Inkwell
//
//  Manages an occasional in-app tip prompt asking the user to support
//  development via Ko-fi. The prompt appears after a few app launches
//  and is suppressed for a cooldown period after dismissal.
//

import Foundation
import SwiftUI
import Observation

@MainActor
@Observable
final class TipPromptManager {
    static let shared = TipPromptManager()

    private let defaults = UserDefaults.standard
    private let launchCountKey = "tipPrompt.launchCount"
    private let lastShownKey = "tipPrompt.lastShownDate"

    private init() {}

    var shouldShowTip: Bool {
        let count = launchCount
        // TipPromptPolicy.shouldShowTip expects millisecond epoch values (it
        // divides the difference by 1000*60*60*24 to get days) — passing
        // timeIntervalSince1970 (seconds) directly here made the computed
        // day-diff 1000x too small, so the 7-day cooldown effectively never
        // re-triggered once it had fired once.
        let lastShown = lastShownDate.map { $0.timeIntervalSince1970 * 1000 } ?? -1
        let now = Date().timeIntervalSince1970 * 1000

        return Inkwell.shouldShowTip(
            launchCount: count,
            lastShownEpochMillis: Int64(lastShown),
            nowEpochMillis: Int64(now)
        )
    }

    func recordLaunch() {
        let current = launchCount
        defaults.set(current + 1, forKey: launchCountKey)
    }

    func markShown() {
        defaults.set(Date(), forKey: lastShownKey)
    }

    // MARK: - Private

    private var launchCount: Int {
        defaults.integer(forKey: launchCountKey)
    }

    private var lastShownDate: Date? {
        defaults.object(forKey: lastShownKey) as? Date
    }
}
